### ISSUE TITLE
"Error: Required type param..."  with Docker containers

### ISSUE THREAD

**Issue by jeff-kilbride** [moviemasher/angular-moviemasher#22]
Hi,

I pulled the docker containers from Docker Hub and started them with the `docker-compose.yml` file you have in the `config/docker/production` directory. They started fine and I was able to view the UI at `localhost:8080`. However, I'm getting the same error as referenced in #15 when I drag and drop a video to the UI. Attached is a screenshot.

<img width="946" alt="screen shot 2017-03-14 at 00 02 28" src="https://cloud.githubusercontent.com/assets/14877960/23889303/9cac4140-0849-11e7-9df0-e0053e7ae502.png">

I traced it to the `/angular-moviemasher/app/php/import_upload.php` script (line 37). To debug, I added the following `error_log` line to the containing block:

```php
if (! $err) { // make sure required parameters have been set
  error_log('$_REQUEST: ' . json_encode($_REQUEST));
  $id = (empty($_REQUEST['id']) ? '' : $_REQUEST['id']);
  $uid = auth_userid();
  $type = (empty($_REQUEST['type']) ? '' : $_REQUEST['type']);
  $extension = (empty($_REQUEST['extension']) ? '' : $_REQUEST['extension']);
  if (! $uid) $err = 'Required uid parameters omitted';
  if (! $id) $err = 'Required id parameters omitted';
  if (! $extension) $err = 'Required extension parameters omitted';
  if (! $type) $err = 'Required type parameters omitted';
}
```
I'm getting the following in the apache logs:
```
angular_moviemasher_1  | [Tue Mar 14 06:59:33.239502 2017] [php7:notice] [pid 58] [client 172.19.0.1:35482] $_REQUEST: [], referer: http://localhost:8080/angular-moviemasher/app/
```
Notice the `$_REQUEST` variable is being output as an empty array.

I'm wondering if this has something to do with the size of the file I'm trying to upload. I am trying to customize the `php.ini` settings to allow for bigger file uploads -- 500M instead of the 200M you have configured by default. To do so, I created my own `php.ini` fragment, based on the one you have in the `config/docker/php.ini` file, and added it to my `docker-compose.yml`:

```yml
version: '2'
services:
  angular_moviemasher:
    image: moviemasher/angular-moviemasher:latest
    ports:
      - "8080:80"
    volumes:
      - ./php.ini:/var/www/html/angular-moviemasher/config/docker/php.ini

  moviemasher_rb:
    image: moviemasher/moviemasher.rb:latest
    volumes_from:
      - angular_moviemasher
    command: process_loop
    tty: true
```
My `php.ini` is as follows:
```
include_path = ".:/var/www/config"
date.timezone = UTC
upload_max_filesize = 500M
post_max_size = 500M
max_input_time = 1200
```
The file that produces the error is 276MB. When I upload a significantly smaller file, 35MB, everything works. Here is the log entry from the smaller file:
```
angular_moviemasher_1  | [Tue Mar 14 07:20:55.707947 2017] [php7:notice] [pid 59] [client 172.19.0.1:35484] $_REQUEST: {"id":"5B44EDBD-B895-475D-A660-681B0D976D7C","type":"video","extension":"mp4"}, referer: http://localhost:8080/angular-moviemasher/app/
```
Obviously, this time the `$_REQUEST` variable is not empty. If size is the issue, there should probably be a better error message than the current one.

So, a couple of questions...  How do I correctly configure `php.ini` to handle bigger uploads? And, is there some kind of user guide for the UI with basic explanations for the different functions / options?

**Comment by syntropo** [comment]
Hmmm... I believe in normal environments we WOULD see a better error surfacing, since we check for size problems specifically. I'm guessing this is some weirdness introduced by Docker - perhaps some sort of limit on how much data can be pushed through to a container. It's probably different for different versions of Docker too. 

You could perhaps test this by setting PHP's limits quite low and uploading a file that exceeds them. I this produces the expected errors, then Docker is likely keeping portions of the request from coming through. 

There is no specific documentation for the PHP code, beyond what's in the code itself. The options in the config file are just documented with comments above.

**Comment by jeff-kilbride** [comment]
Thanks for the quick reply!

It doesn't look like the changes I'm making to the `php.ini` file are having any effect. I put a simple `info.php` script in the `app` directory with:

```php
<?php phpinfo(); ?>
```
No matter what values I put in `php.ini`, the settings are always:
```
max_file_uploads = 20
max_input_time = 600
post_max_size = 200M
```
I think I was originally confusing `max_file_uploads` with `upload_max_filesize`. Anyway, when I try to change the values, nothing happens. I lowered `post_max_size` to 20M, but was still able to successfully upload a 35MB file. So, I'm not sure what's happening. 

How do I configure PHP inside the Docker container? I can't find any other `php.ini` files, except the one in `config/docker`.

**Comment by syntropo** [comment]
Well, that file is only copied into the container when it's built - from the Dockerfile:

`COPY config/docker/php.ini /usr/local/etc/php/conf.d/moviemasher.ini`

If you want it to be mounted live, you could add a line to the volumes array in docker-compose. Alternatively, you could use exec to edit the file within a running container. Either way, Apache needs to be restarted before changes to PHP INI files take effect.

**Comment by jeff-kilbride** [comment]
Ok. I didn't see that line in your Dockerfile. I was trying to replace `config/docker/php.ini`:
```
    volumes:
      - ./php.ini:/var/www/html/angular-moviemasher/config/docker/php.ini
```
I've changed my volume spec and now I'm able to change the settings for PHP. The "Error: required type param..." seems to be coming from reaching PHP's upload file limit. With these settings:
```
include_path = ".:/var/www/config"
date.timezone = UTC
upload_max_filesize = 10M
post_max_size = 10M
max_input_time = 600
```
I get the error when I try to upload a 35MB file. For whatever reason, the `$_REQUEST` variable isn't being set. Similarly, when I raise `post_max_size` and `upload_max_filesize` to 500MB, I am able to upload the 276MB file I was having problems with earlier.

**Comment by syntropo** [comment]
Uh, I think the volumes element would be:
`../php.ini:/usr/local/etc/php/conf.d/moviemasher.ini`

This replaces the moviemasher.ini file in PHP's configuration directory with a link to config/docker/php.ini (paths are relative to the docker-compose file). And this change is only applicable to newly started containers. 

It's not been my experience that PHP wipes the $_REQUEST when there was a problem uploading a file. Typically it adds entries to $_FILES for each one, and populates the `error` key if there was a problem with any or all of them.

**Comment by jeff-kilbride** [comment]
Sorry if I'm not being clear. This is my `docker-compose.yml` that is working:
```yml
version: '2'
services:
  angular_moviemasher:
    image: moviemasher/angular-moviemasher:latest
    ports:
      - "8080:80"
    volumes:
      - ./php.ini:/usr/local/etc/php/conf.d/moviemasher.ini

  moviemasher_rb:
    image: moviemasher/moviemasher.rb:latest
    volumes_from:
      - angular_moviemasher
    command: process_loop
    tty: true
```
My custom `php.ini` is in the same directory as my `docker-compose.yml` file:
```ini
include_path = ".:/var/www/config"
date.timezone = UTC
upload_max_filesize = 500M
post_max_size = 500M
max_input_time = 1200
```
I use Docker extensively in my dev environment. The `volumes` section in my last post was just an example of how I had gotten it wrong. I didn't realize you copied and renamed your `config/docker/php.ini` file to `/usr/local/etc/php/conf.d/moviemasher.ini`. I was initially looking for another `php.ini` file in the system, but `config/docker/php.ini` was the only one I could find. So, I tried injecting my custom one over it. Obviously, it didn't work. My mistake.

I'm not sure how it works in other environments, because I've only tested the Docker containers. In my experience with these containers, if you try to upload a file bigger than the PHP settings allow, the `$_REQUEST` variable in `import_upload.php` is empty. It's very easy to reproduce. Start the containers and try uploading a file bigger than 200MB (the default). You'll get the red error message you see in my first post. This may be the same problem the user from #15 was experiencing. 

Since I've solved this issue using my own `php.ini` settings, feel free to close this. Just thought you (and possibly others who run into this problem) would like to know where it's occurring and how to fix it. Whether you dig any deeper into why the `$_REQUEST` variable is empty when run in Docker is up to you. My suggestion would be to put something in the Docker README that says the default max upload size is 200MB. The current error is fairly confusing. I'm familiar with Docker and PHP, so I was able to trace it on my own. Others who just want to test your software using the containers may not be able to figure it out.

Thanks!

**Comment by syntropo** [comment]
Thanks for pointing the Docker issue out. Seems I would be getting more bug reports if this weren't working correctly on the AMI and other environments, but I will do some testing. 

And yes, the error is absolutely confusing - I will test out in Docker too. Is $_FILES also empty? How on earth can you trap for error conditions without it? As you can see in import_upload.php, we're going out of way to report every possible problem in plain english. I find it funny that Docker is killing the request, but not the cookies - note that you made it past the authentication step!

And sorry for the confusion about the ini file - I thought you were just customizing the one in the directory above.

**Comment by TharakaRumeshJeewantha** [comment]
Hello

I downloaded Angular-MovieMasher on GitHub and I tried to run it on my WAMP and XAMPP server.
I want to remove docker part on this project and then it run on my servers. 

![aaa](https://user-images.githubusercontent.com/25604416/49323431-84883b00-f541-11e8-9ba9-5baea9c71b4f.PNG)
But I can't uploaded the video. please help me to config and fix this problem.

Thank you

**Comment by syntropo** [comment]
You'll want to read through the 'Installation' and 'Install on your web server' sections of the README for this project. Please understand that the moviemasher.rb must also be utilized, but should be installed on a separate server since it is so resource intensive.
